packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts

```diff
@@ -0,0 +1,147 @@
+    public errorType: string = "unreferencedObjectUsedAfterGarbageCollected";
+}
+
+/**
+ * This class encapsulates the logic around what to do when a SweepReady object is used.
+ * There are several tactics we plan to use in Dogfood environments to aid diagnosis of these cases:
+ *  - Closing the main container when either the main or summarizer client detects this kind of violation
+ *    (via sweepReadyUsageDetectionSetting above)
+ *  - Throttling the frequency of these crashes via a "Blackout Period" per container per device
+ *   (via blackoutPeriodDaysKey above.  Uses localStorage and closuresMapLocalStorageKey to implement this behavior)
+ */
+export class SweepReadyUsageDetectionHandler {
+    private readonly localStorage: Pick<Storage, "getItem" | "setItem">;
+
+    constructor(
+        private readonly uniqueContainerKey: string,
+        private readonly mc: MonitoringContext,
+        private readonly closeFn: (error?: ICriticalContainerError) => void,
+        localStorageOverride?: Pick<Storage, "getItem" | "setItem">,
+    ) {
+        const noopStorage = { getItem: () => null, setItem: () => {} };
+        if (localStorageOverride !== undefined) {
+            this.localStorage = localStorageOverride;
+        } else {
+            try {
+                // localStorage is not defined in Node environment so this throws
+                this.localStorage = localStorage ?? noopStorage;
+            } catch (error) {
+                this.localStorage = noopStorage;
+            }
+        }
+
+        if (this.localStorage === noopStorage) {
+            // This means the Blackout Period logic will not work.
+            this.mc.logger.sendTelemetryEvent({ eventName: "SweepReadyUsageDetectionHandlerNoOpStorage" });
+        }
+    }
+
+    /**
+      * If SweepReady Usage Detection is enabled, close the main container.
+      * If the "Blackout Period" is set, don't close the container more than once in that period.
+      *
+      * Once Sweep is fully implemented, this will be removed since the objects will be gone
+      * and errors will arise elsewhere in the runtime
+     */
+    public usageDetectedInMainContainer(errorProps: ITelemetryProperties) {
+        if (!sweepReadyUsageDetectionSetting.read(this.mc.config).mainContainer) {
+            return;
+        }
+
+        // Default stance is we close every time - this reflects the severity of SweepReady Object Usage.
+        // However, we may choose to "throttle" the closures by setting the BlackoutPeriodDays setting,
+        // which will only allow the container to close once during that period, to avoid locking users out.
+        let shouldClose: boolean = true;
+        let pastClosuresMap: Record<string, { lastCloseTime: number; } | undefined> = {};
+        let lastCloseTime: number | undefined;
+        const blackoutPeriodDays = this.mc.config.getNumber(blackoutPeriodDaysKey);
+        if (blackoutPeriodDays !== undefined) {
```

Comment:
Should you skip this if localStorage is noopStorage?

Reply:
It works as-is (there's test coverage), that's an unexpected case, so not wanting to call it out.